File: src/docker/ready.rs
use anyhow::{bail, Context, Result};
use backon::{ExponentialBuilder, Retryable};
use bollard::query_parameters::LogsOptions;
use bollard::Docker;
use futures_util::StreamExt;
use std::time::Duration;

use crate::config::model::ReadyCheck;
use crate::docker::exec::exec_in_container;

/// Run a ready check for a container, retrying with exponential backoff.
///
/// Dispatches to the appropriate strategy based on the ReadyCheck variant.
pub async fn run_ready_check(
    docker: &Docker,
    container_id: &str,
    check: &ReadyCheck,
    host_port: Option<u16>,
    docker_name: &str,
) -> Result<()> {
    let total_timeout = Duration::from_secs(check.timeout_secs());

    match check {
        ReadyCheck::Log { pattern, .. } => {
            run_log_check(docker, container_id, pattern, total_timeout, docker_name).await
        }
        _ => {
            let docker = docker.clone();
            let container_id = container_id.to_string();
            let check = check.clone();
            let docker_name = docker_name.to_string();

            let result = tokio::time::timeout(total_timeout, async {
                (|| async { run_single_check(&docker, &container_id, &check, host_port).await })
                    .retry(
                        ExponentialBuilder::default()
                            .with_min_delay(Duration::from_millis(250))
                            .with_max_delay(Duration::from_secs(3))
                            .with_max_times(200)
                            .with_jitter(),
                    )
                    .notify(|err: &anyhow::Error, dur: Duration| {
                        tracing::debug!(
                            docker = %docker_name,
                            "ready check failed: {}, retrying in {:?}",
                            err,
                            dur
                        );
                    })
                    .await
            })
            .await;

            match result {
                Ok(Ok(())) => Ok(()),
                Ok(Err(e)) => Err(e),
                Err(_) => bail!(
                    "ready check for '{}' timed out after {:?}",
                    docker_name,
                    total_timeout
                ),
            }
        }
    }
}

/// Run a single (non-retrying) check based on the strategy.
async fn run_single_check(
    docker: &Docker,
    container_id: &str,
    check: &ReadyCheck,
    host_port: Option<u16>,
) -> Result<()> {
    match check {
        ReadyCheck::PgIsReady { .. } => {
            let cmd = vec![
                "pg_isready".to_string(),
                "-h".to_string(),
                "localhost".to_string(),
                "-q".to_string(),
                "-t".to_string(),
                "2".to_string(),
            ];
            let (exit_code, _) = exec_in_container(docker, container_id, cmd).await?;
            if exit_code != 0 {
                bail!("pg_isready returned exit code {}", exit_code);
            }
            Ok(())
        }
        ReadyCheck::Cmd { command, expect, .. } => {
            crate::docker::exec::exec_ready_check(docker, container_id, command, expect.as_deref())
                .await
        }
        ReadyCheck::Http { url, .. } => {
            let client = reqwest::Client::builder()
                .timeout(Duration::from_secs(2))
                .build()
                .context("building HTTP client")?;
            let response = client.get(url).send().await.context("HTTP ready check")?;
            if !response.status().is_success() {
                bail!("HTTP ready check returned status {}", response.status());
            }
            Ok(())
        }
        ReadyCheck::Tcp { .. } => {
            let port = host_port.context("TCP ready check requires a port")?;
            tokio::time::timeout(
                Duration::from_secs(2),
                tokio::net::TcpStream::connect(format!("127.0.0.1:{}", port)),
            )
            .await
            .context("TCP connect timed out")?
            .context("TCP connect failed")?;
            Ok(())
        }
        ReadyCheck::Log { .. } => {
            unreachable!("log check handled separately")
        }
    }
}

/// Run a log-based ready check by streaming container logs and scanning for
/// a pattern match.
async fn run_log_check(
    docker: &Docker,
    container_id: &str,
    pattern: &str,
    timeout: Duration,
    docker_name: &str,
) -> Result<()> {
    let options = LogsOptions {
        follow: true,
        stdout: true,
        stderr: true,
        tail: "all".to_string(),
        ..Default::default()
    };

    let mut stream = docker.logs(container_id, Some(options));

    let result = tokio::time::timeout(timeout, async {
        while let Some(msg) = stream.next().await {
            match msg {
                Ok(log) => {
                    let text = log.to_string();
                    if text.contains(pattern) {
                        return Ok(());
                    }
                }
                Err(e) => {
                    tracing::warn!(docker = %docker_name, "log stream error: {}", e);
                }
            }
        }
        bail!("log stream ended without finding pattern '{}'", pattern)
    })
    .await;

    match result {
        Ok(Ok(())) => Ok(()),
        Ok(Err(e)) => Err(e),
        Err(_) => bail!(
            "log ready check for '{}' timed out after {:?} (pattern: '{}')",
            docker_name,
            timeout,
            pattern
        ),
    }
}

File: internal/docker/ready.go
package docker

import (
	"context"
	"fmt"
	"io"
	"net"
	"net/http"
	"strings"
	"time"

	dockerclient "github.com/docker/docker/client"
	"github.com/docker/docker/api/types/container"

	"github.com/steveyackey/devrig/internal/config"
)

// RunReadyCheck polls until the container passes its ready check or the
// configured timeout expires.
func RunReadyCheck(
	ctx context.Context,
	cli *dockerclient.Client,
	containerID string,
	check *config.ReadyCheck,
	hostPort *uint16,
	name string,
) error {
	timeout := time.Duration(check.TimeoutSecs()) * time.Second
	if timeout == 0 {
		timeout = 60 * time.Second
	}

	if check.Type == "log" {
		return waitForLogPattern(ctx, cli, containerID, check.Match, timeout, name)
	}

	deadline := time.Now().Add(timeout)
	delay := 250 * time.Millisecond
	maxDelay := 3 * time.Second

	for {
		err := runSingleCheck(ctx, cli, containerID, check, hostPort)
		if err == nil {
			return nil
		}
		if time.Now().After(deadline) {
			return fmt.Errorf("ready check for %q timed out after %s: %v", name, timeout, err)
		}
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(delay):
		}
		delay = min(delay*2, maxDelay)
	}
}

func runSingleCheck(ctx context.Context, cli *dockerclient.Client, containerID string, check *config.ReadyCheck, hostPort *uint16) error {
	switch check.Type {
	case "pg_isready":
		code, _, err := execInContainer(ctx, cli, containerID, []string{"pg_isready", "-h", "localhost", "-q", "-t", "2"})
		if err != nil {
			return err
		}
		if code != 0 {
			return fmt.Errorf("pg_isready returned exit code %d", code)
		}
		return nil

	case "cmd":
		return ExecReadyCheck(ctx, cli, containerID, check.Command, check.Expect)

	case "http":
		httpClient := &http.Client{Timeout: 2 * time.Second}
		resp, err := httpClient.Get(check.URL)
		if err != nil {
			return err
		}
		_ = resp.Body.Close()
		if resp.StatusCode < 200 || resp.StatusCode >= 300 {
			return fmt.Errorf("HTTP ready check returned status %d", resp.StatusCode)
		}
		return nil

	case "tcp":
		if hostPort == nil {
			return fmt.Errorf("TCP ready check requires a port")
		}
		conn, err := net.DialTimeout("tcp", fmt.Sprintf("127.0.0.1:%d", *hostPort), 2*time.Second)
		if err != nil {
			return err
		}
		_ = conn.Close()
		return nil

	default:
		return fmt.Errorf("unknown ready check type %q", check.Type)
	}
}

func waitForLogPattern(ctx context.Context, cli *dockerclient.Client, containerID, pattern string, timeout time.Duration, name string) error {
	timeoutCtx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()

	rc, err := cli.ContainerLogs(timeoutCtx, containerID, container.LogsOptions{
		ShowStdout: true,
		ShowStderr: true,
		Follow:     true,
		Tail:       "all",
	})
	if err != nil {
		return fmt.Errorf("streaming logs for %s: %w", name, err)
	}
	defer rc.Close()

	buf := make([]byte, 4096)
	var accumulated strings.Builder
	for {
		n, err := rc.Read(buf)
		if n > 0 {
			accumulated.Write(buf[:n])
			if strings.Contains(accumulated.String(), pattern) {
				return nil
			}
		}
		if err != nil {
			if err == io.EOF {
				break
			}
			if timeoutCtx.Err() != nil {
				break
			}
		}
	}
	return fmt.Errorf("log ready check for %q timed out after %s (pattern: %q)", name, timeout, pattern)
}

func min(a, b time.Duration) time.Duration {
	if a < b {
		return a
	}
	return b
}
